### PR TITLE
Updated getTicks - Condition codes and multiple fields/event types

### PR DESCRIPTION
**Pull request by forkandspoon**
Updated getTicks to support condition code overwrite through a bool toggle, default is `FALSE`.

Multiple field types supported i.e. `eventType = c("TRADE","BEST_ASK","BEST_BID")`. Default is `TRADE`, output in long format.

e.g. `getTicks(con, "XUA Index", eventType = c("BEST_ASK", "BEST_BID"), startDateTime = as.POSIXct("2015-07-07 10:00"), endDateTime = as.POSIXct("2015-07-07 12:00"), setCondCodes = TRUE)`

Doc is updated in getTicks.R

### CODE REVIEW

File: R/getTicks.R [hunk 1]
```diff
@@ -23,18 +26,19 @@
 ##' an object of the type selected in \code{returnAs}.
 ##' @author Dirk Eddelbuettel
 getTicks <- function(security,
-                     eventType = "TRADE",
+                     eventType = c("TRADE"),
```
Comment (eddelbuettel) [inline R/getTicks.R]: This doesn't need `c("TRADE")` as `"TRADE"` is still the same (but I like you made the argument a vector).

File: src/getTicks.cpp [hunk 2]
```diff
@@ -43,6 +43,8 @@
  * IN THE SOFTWARE.
  */
 
+#include <vector>
+#include <string>
```
Comment (eddelbuettel) [inline src/getTicks.cpp]: I think we get both of those via Rcpp.h as well, but doesn't hurt to be explicit.

**Comment by eddelbuettel** [comment]
Looks good (apart from the mini-notes I added above).  Appreciate the clean indentation and general adherence to our style.

I'll fold this in.

**Comment by eddelbuettel** [comment]
@forkandspoon Would you mind sharing your full name and id, though?  Feel free to email me privately. We do have you gmail address from the commit log, but it would be nice to have a full name too. Thanks.

**Comment by forkandspoon** [comment]
Thanks, I've added my name to my profile, please let me know if there any issues.

**Comment by eddelbuettel** [comment]
@forkandspoon @armstrtw @johnlaing 

I need to do some work on this.   With this pull request in, we now get 

```r
R> dat <- getTicks(sym, c("TRADE", "BEST_BID", "BEST_ASK"), Sys.time()-60*60, Sys.time())
R> head(dat)
                    type       value   size  conditionCode
2015-07-17 11:20:04 "BEST_BID" "25.10" "  6" ""           
2015-07-17 11:20:04 "BEST_BID" "25.11" "  1" ""           
2015-07-17 11:20:04 "BEST_ASK" "25.14" "  6" ""           
2015-07-17 11:20:04 "BEST_BID" "25.12" "  1" ""           
2015-07-17 11:20:30 "BEST_BID" "25.11" "  1" ""           
2015-07-17 11:20:30 "TRADE"    "25.12" "100" ""           
R> 
```

which is _character matrices_.  Ewwwww.

Before and in 0.2.2.5 when just `TRADE` was supported we get
```r
R> head(dat1)
                    value size
2015-07-17 11:39:27 25.12  100
R> 
```

which is of type `xts` for me (by choice) using numeric matrices.  

I want to preserve that.   Comments?

**Comment by n-e-w** [comment]
Do you mean in terms of whether it is best to have as an `xts` or as another type? Personally, I think most users of this package would want the output in xts format. But, as we know, `xts` doesn't allow characters...what to do...what to do...

Here's a quick kludge to kick off discussion

First up, I don't get character matrices. 

With 0.2.2.6 installed and using SPY US Equity, i get:
```r
> dat <- getTicks("SPY US Equity",
 c("TRADE", "BEST_BID", "BEST_ASK"), Sys.time()-60*60, Sys.time())

> str(dat)

'data.frame':	105 obs. of  5 variables:
 $ times        : POSIXct, format: "2015-07-17 21:58:53" "2015-07-17 21:58:53" "2015-07-17 21:58:53" "2015-07-17 21:58:53" ...
 $ type         : Factor w/ 3 levels "BEST_ASK","BEST_BID",..: 2 1 2 1 2 1 2 1 2 1 ...
 $ value        : num  212 212 212 212 212 ...
 $ size         : num  13 2 5 2 2 2 7 2 7 1 ...
 $ conditionCode: Factor w/ 1 level "": 1 1 1 1 1 1 1 1 1 1 ...
```

Be that as it may, I think one solution is to convert the data.frame into an xts in the format most traders / quants are used to -- columns of bid/ask/trade with corresponding size (a little like in <QRM> on the terminal)

In case character matrices are returned and you want to work with a lot of ticks, I've used the `fasttime` package to make the timestamp conversion as fast as possible.

```r
#install.packages('fasttime',repos='http://www.rforge.net/')

convertBidAskTrade <- function(dat){


library(fasttime)

#Subset into bid / ask/ trade data frames (used subset to make it explicit for example)
  
bbid <- subset(dat,dat$type == "BEST_BID")
bask <- subset(dat,dat$type == "BEST_ASK")
btrade <- subset(dat,dat$type == "TRADE")

# convert each frame to an xts. I've used fastPOSIXct from `fasttime` 
# for converting very large series -- much faster.
# Less relevant if the dataframe already has POSIXct timestamps instead of character

bbid_xts <- xts(bbid[,3:4],fastPOSIXct(bbid[,1],tz="UTC"))
colnames(bbid_xts) <- c("BEST_BID","BEST_BID_SZ")

bask_xts <- xts(bask[,3:4],fastPOSIXct(bask[,1],tz="UTC"))
colnames(bask_xts) <- c("BEST_ASK","BEST_ASK_SZ")

btrade_xts <- xts(btrade[,3:4],fastPOSIXct(btrade[,1],tz="UTC"))
colnames(btrade_xts) <- c("TRADE","TRADE_SZ")

#merge these together
frm <- merge(bbid_xts,bask_xts,btrade_xts)

#Use na.locf to carry over last bid and ask data to 'cover' when a trade occurs
frm[,1:4] <- na.locf(frm[,1:4])
return(frm)
}

# show the tail of the new xts to show a trade (first row)
> tail(x,32)
                    BEST_BID BEST_BID_SZ BEST_ASK BEST_ASK_SZ  TRADE TRADE_SZ
2015-07-17 22:40:03   212.49          22   212.51           5 212.48        0
2015-07-17 22:40:27   212.49          22   212.50          10     NA       NA
2015-07-17 22:42:15   212.49          18   212.50          10     NA       NA
2015-07-17 22:42:28   212.49          13   212.50          10     NA       NA
2015-07-17 22:45:47   212.49           9   212.50          10     NA       NA
2015-07-17 22:53:25   212.49           4   212.50          10     NA       NA
2015-07-17 22:54:45   212.49           4   212.49           1     NA       NA
2015-07-17 22:54:45   212.48           5   212.49           1     NA       NA
2015-07-17 22:54:50   212.47          21   212.48           1     NA       NA
2015-07-17 22:54:56   212.47          21   212.50          10     NA       NA
2015-07-17 22:55:16   212.47           7   212.50          10     NA       NA
2015-07-17 22:55:16   212.47           6   212.50          10     NA       NA
2015-07-17 22:55:16   212.47           4   212.50          10     NA       NA
2015-07-17 22:55:39   212.47           4   212.49          20     NA       NA
2015-07-17 22:56:11   212.45           2   212.49          20     NA       NA
2015-07-17 22:56:11   212.45           1   212.49          20     NA       NA
2015-07-17 22:56:11   212.43           2   212.49          20     NA       NA
2015-07-17 22:56:12   212.43           2   212.45           1     NA       NA
2015-07-17 22:56:23   212.43           2   212.49          20     NA       NA
2015-07-17 22:56:23   212.45           3   212.49          20     NA       NA
2015-07-17 22:56:29   212.45           1   212.49          20     NA       NA
2015-07-17 22:56:33   212.43           2   212.49          20     NA       NA
2015-07-17 22:56:33   212.44           1   212.49          20     NA       NA
2015-07-17 22:56:35   212.43           2   212.49          20     NA       NA
2015-07-17 22:56:37   212.44           1   212.49          20     NA       NA
2015-07-17 22:56:44   212.43           2   212.49          20     NA       NA
2015-07-17 22:56:48   212.43           2   212.49          21     NA       NA
2015-07-17 22:56:52   212.42           1   212.49          21     NA       NA
2015-07-17 22:56:52   212.42           1   212.48           2     NA       NA
2015-07-17 22:57:03   212.43          25   212.48           2     NA       NA
2015-07-17 22:57:26   212.43          25   212.47           2     NA       NA
2015-07-17 22:57:37   212.43          25   212.46           2     NA       NA

```

Just a thought. Feedback appreciated.

**Comment by eddelbuettel** [comment]
Something like that is what I had in mind too.   Quick comments, though:
* fasttime, as much as I like it and use it, is not needed here: Note that you already reported ` $ times        : POSIXct, format: "2015-07-17 21:58:53" ...`
* rearringing the data into ntypes*2 columns can be done at the C++ level, at which point we are back to returning an N * (ntypes*2) matrix

**Comment by armstrtw** [comment]
Not all users prefer xts.  I think if the return value has chars and we
wish to return them,then the only choice is a dataframe.

**Comment by eddelbuettel** [comment]
I think we can do what we did before: return a rectangular chunk with an (optional) converter to xts, fts, zoo, ...

**Comment by eddelbuettel** [comment]
It's tricky due to the asynchronous timestamps.  Maybe the script by @n-e-w above is the best we can do:  Loop over the different 'types' and return the values and sizes.

**Comment by forkandspoon** [comment]
I have been doing something similar to the above, converting the output into a wide format. The condition chars weren't included in the analysis, rather just providing some additional colour before. All for the new format though!

**Pull request merged**